docs/source/extension/extension_points.rst

```diff
@@ -721,6 +723,117 @@ When the ``labStatus`` busy state changes, we update the text content of the
+       );
+       
+       toolbarRegistry.registerFactory<NotebookPanel>(
+         FACTORY,
+         'kernelStatus',
+         panel => Toolbar.createKernelStatusItem(panel.sessionContext, translator)
+       );
+       // etc... 
+     
+       if (settingRegistry) {
+         // Create the factory
+         toolbarFactory = createToolbarFactory(
+           toolbarRegistry,
+           settingRegistry,
+           // Factory name
+           FACTORY,
+           // Setting id in which the toolbar items are defined
+           '@jupyterlab/notebook-extension:panel',
+           translator
+         );
+       }
+     }
+   
+     const factory = new NotebookWidgetFactory({
+       name: FACTORY,
+       fileTypes: ['notebook'],
+       modelName: 'notebook',
+       defaultFor: ['notebook'],
+       // ...
+       toolbarFactory,
+       translator: translator
+     });
+     app.docRegistry.addWidgetFactory(factory);
+
+The registry ``registerFactory`` method allows an extension to provide special widget for a unique pair 
+(factory name, toolbar item name). Then the helper ``createToolbarFactory`` can be used to extract the
+toolbar definition from the settings and build the factory to pass to the widget factory.
+
+The default toolbar items can be defined across multiple extensions by providing an entry in the ``"jupyter.lab.toolbars"``
+mapping. For example for the notebook panel:
+
+.. code:: json
+ 
+   "jupyter.lab.toolbars": {
+     "Notebook": [
+       { "name": "save", "rank": 10 },
+       { "name": "insert", "command": "notebook:insert-cell-below", "rank": 20 },
+       { "name": "cut", "command": "notebook:cut-cell", "rank": 21 },
+       { "name": "copy", "command": "notebook:copy-cell", "rank": 22 },
+       { "name": "paste", "command": "notebook:paste-cell-below", "rank": 23 },
+       { "name": "run", "command": "runmenu:run", "rank": 30 },
+       { "name": "interrupt", "command": "kernelmenu:interrupt", "rank": 31 },
+       { "name": "restart", "command": "kernelmenu:restart", "rank": 32 },
+       {
+         "name": "restart-and-run",
+         "command": "runmenu:restart-and-run-all",
+         "rank": 33
+       },
+       { "name": "cellType", "rank": 40 },
+       { "name": "spacer", "type": "spacer", "rank": 100 },
+       { "name": "kernelName", "rank": 1000 },
+       { "name": "kernelStatus", "rank": 1001 }
+     ]
+   },
+   "jupyter.lab.transform": true,
+   "properties": {
+     "toolbar": {
+       "title": "Notebook panel toolbar items",
+       "items": {
+         "$ref": "#/definitions/toolbarItem"
+       },
+       "type": "array",
+       "default": []
+     }
+   }
+
+The settings registry will gather those definitions within the provided setting plugin id
+under the ``toolbar`` property. That list will be used to create the toolbar. The user
+can customize it by adding new items or overriding existing one (like providing a different
+rank or add ``"disabled": true`` to remove the item).
```

Comment:
This is what I understood, I hope it makes it clearer. Please edit to correct if not accurate or apply stylistic changes as suits you.

```suggestion
The settings registry will merge those definitions from settings schema with any
user-provided overrides (customization) transparently and save them under the
``toolbar`` property in the final settings object. The ``toolbar`` list will be used to
create the toolbar. Both the source settings schema and the final settings object
are identified by the plugin ID passed to ``createToolbarFactory``. The user can
customize the toolbar by adding new items or overriding existing ones (like
providing a different rank or adding ``"disabled": true`` to remove the item).
```

Reply:
Thanks a lot for taking the time to improve the documentation